duckduckgo/Android
# Feature/improve tab ux

Task/Issue URL: https://app.asana.com/0/1126768040926570/1126588912559580
Tech Design URLs: 

- https://app.asana.com/0/1126768040926570/1127542612010527
- https://app.asana.com/0/1126768040926570/1130402521558417
- https://app.asana.com/0/1126768040926570/1130402521558423
- https://app.asana.com/0/1126768040926570/1130402521558416
- https://app.asana.com/0/1126768040926570/1130402521558415

**Description**:
Adds new tab switcher UX. There's quite a few changes here, but the highlights are:
1. Grid view
1. WebView preview images
1. Swipe to close tabs

ℹ️We will be A/B testing this change alongside the existing tab switcher implementation. To test the new changes, configure your variant to be `mx`.

ℹ️When generating the webview previews, it can take ~1s for a change you see on the screen to be available to the webview preview API. Presumably this is because there is a mechanism in the view to show the absolute latest on screen immediately, but through the API, there might be a slight delay in when that data is available. As such, empirically, the preview generated will usually reflect what was visible on the WebView about ~1s ago.

**Steps to test this PR**:
1. From a blank tab state, visit the tab switcher. verify there is no preview shown (blank area in card)
1. Type a search query, and let the SERP load. Visit tab switcher; verify the SERP preview is shown, and the title of the card shows the query (and not the actual page title, ie, it'll have no "at DuckDuckGo")
1. Visit a webpage, then visit the tab switcher; verify the webview preview is shown, and that the title of the card shows the title of the website
1. Add multiple tabs. Verify that no matter which tab is currently selected, when you launch the tab switcher view the currently selected tab is always on screen (even if it was at the bottom of the tab list)
1. Swipe to delete a tab. Verify it works as you'd expect.
1. Click the `X` to delete a tab; Verify it still works as you'd expect
1. From a webpage (e.g., cnn.com) choose to open a link in a background tab. Visit tab switcher. Verify the favicon shows an "unread" indicator, and that there is no webview preview available

For the following tests, you'll need to access the device file system (so use an emulator or rooted device).

ℹ️The webview previews are stored on disk in the `data/data/com.duckduckgo.mobile.android/cache/tabPreviews` directory

1. Visit a webpage; verify you can see an image in the tab previews directory for
1. Navigate around to different pages; verify there is only a current tab preview available and old ones are deleted as you navigate
1. Close the tab; verify there are no tab previews for that tab
1. Open multiple tabs; use the "close all tabs" option from the tab switcher; verify that there are no tab previews left on disk
1. Choose to "automatically clear tabs" from the app settings. Open a few tabs, and then let the auto data clearer kick in. Verify no tab previews left on disk.
1. Choose to "automatically clear tabs and data" from the app settings. Open a few tabs, and then let the auto data clearer kick in. Verify no tab previews left on disk.


Database migration tests
1. Check everything works from a fresh install
1. Uninstall; install `develop` and open a few tabs. Perform an update to this version, and check everything works as expected. No tab previews will be available for existing tabs until you reload them.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
